packages/hydrogen/src/pagination/Pagination.ts

```diff
@@ -0,0 +1,263 @@
+import {createElement, useEffect, useMemo, useState} from 'react';
+import type {Maybe, PageInfo} from '@shopify/hydrogen/storefront-api-types';
+
+import {Link, LinkProps, useNavigation, useLocation} from '@remix-run/react';
+
+type Connection<NodesType> = {
+  nodes: Array<NodesType>;
+  pageInfo: PageInfo;
+};
+
+type PaginationState<NodesType> = {
+  nodes: Array<NodesType>;
+  pageInfo: PageInfo | null;
+};
+
+interface PaginationInfo<NodesType> {
+  /** The paginated array of nodes. You should map over and render this array. */
+  nodes: Array<NodesType>;
+  /** The `<NextLink>` is a helper component that makes it easy to navigate to the next page of paginated data. Alternatively you can build your own `<Link>` component: `<Link to={nextPageUrl} state={state} preventScrollReset />` */
+  NextLink: (props: Omit<LinkProps, 'to'>) => JSX.Element | null;
+  /** The `<PreviousLink>` is a helper component that makes it easy to navigate to the previous page of paginated data. Alternatively you can build your own `<Link>` component: `<Link to={previousPageUrl} state={state} preventScrollReset />` */
+  PreviousLink: (props: Omit<LinkProps, 'to'>) => JSX.Element | null;
+  /** The URL to the previous page of paginated data. Use this prop to build your own `<Link>` component. */
+  previousPageUrl: string;
+  /** The URL to the next page of paginated data. Use this prop to build your own `<Link>` component. */
+  nextPageUrl: string;
+  /** True if the cursor has next paginated data */
+  hasNextPage: boolean;
+  /** True if the cursor has previous paginated data */
+  hasPreviousPage: boolean;
+  /** True if we are in the process of fetching another page of data */
+  isLoading: boolean;
+  /** The `state` property is important to use when building your own `<Link>` component if you want paginated data to continuously append to the page. This means that every time the user clicks "Next page", the next page of data will be apppended inline with the previous page. If you want the whole page to re-render with only the next page results, do not pass the `state` prop to the Remix `<Link>` component. */
+  state: {
+    nodes: Array<NodesType>;
+    pageInfo: {
+      endCursor: Maybe<string> | undefined;
+      startCursor: Maybe<string> | undefined;
+      hasPreviousPage: boolean;
+    };
+  };
+}
+
+type PaginationProps<NodesType> = {
+  /** The response from `storefront.query` for a paginated request. Make sure the query is passed pagination variables and that the query has `pageInfo` with `hasPreviousPage`, `hasNextpage`, `startCursor`, and `endCursor` defined. */
+  connection: Connection<NodesType>;
+  /** A render prop that includes pagination data and helpers. */
+  children: PaginationRenderProp<NodesType>;
+};
+
+type PaginationRenderProp<NodesType> = (
+  props: PaginationInfo<NodesType>,
+) => JSX.Element | null;
+
+export function Pagination<NodesType>({
+  connection,
+  children = () => null,
```

Comment:
Hmm, is this the right type for children? I would expect it to return JSX, right?